### PR TITLE
#166265916 Implement Social media Sign in

### PR DESCRIPTION
**Pull request by cvjude**
**What does this PR do?**
Implement Social media Sign in to the app

**Description of Task to be completed?**
>- Use passport access_token verification
>- Create facebook app, google app, twitter app

**How should this be manually tested?**
Clone the repo and run npm install to download all packages, then use npm run test to test the endpoint

**Any background context you want to provide?**
None

**What are the relevant pivotal tracker stories?**
<a href = "https://www.pivotaltracker.com/story/show/#166265916">#166265916</a>

**Screenshots (if appropriate)**
>Facebook Login
facebook login

>Google login
googlelogin

### CODE REVIEW

File: server/__test__/user.spec.js [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+      console.log(error);
+    }
+  });
+
+  afterAll(async () => {
+    try {
+      await migrateSeeders.down();
+      console.log('data successfully deleted');
+    } catch (error) {
+      console.log(error);
```
[the remark this answers is not in the document]
Comment (chuxmykel) [reply server/__test__/user.spec.js]: Nice work Jude. Please let's use debug rather than console.log

File: server/controllers/userController.js [hunk 2]
```diff
@@ -1,55 +1,56 @@
...
     }
-    
-}
+  }
 
+  static async socialSignin(req, res) {
+    const { userProfile } = req;
+    const token = auth.generateToken({ id: userProfile.id });
+
+    return util.successStatus(res, 201, 'Login successful', {
```
Comment (chuxmykel) [inline server/controllers/userController.js]: 200 status I guess

File: server/db/migrations/20190528151100-create-user.js [hunk 3]
```diff
@@ -10,25 +10,37 @@ export function up(queryInterface, Sequelize) {
...
     },
     email: {
       type: Sequelize.STRING,
-      allowNull: false,
+      allowNull: true
     },
     password: {
       type: Sequelize.STRING,
-      allowNull: false,
+      allowNull: true
```
Comment (chuxmykel) [inline server/db/migrations/20190528151100-create-user.js:25]: Why are we allowing null here?

**Review by chuxmykel — CHANGES_REQUESTED** [review]

File: server/routes/index.js [hunk 4]
```diff
@@ -1,18 +1,20 @@
 import express from 'express';
 import swaggerUI from 'swagger-ui-express';
 import yaml from 'yamljs';
-import userRouter from "./userRoute"
+import userRouter from './userRoute';
+import socialRouter from './socialRoute';
```
Comment (ocranbillions) [inline server/routes/index.js:6]: I was thinking you were going to add all the routes in one file.
Nice work though

File: server/__test__/user.spec.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (cvjude) [reply server/__test__/user.spec.js]: I'll work on that as soon as possible

[the remark this answers is not in the document]
Comment (cvjude) [reply server/__test__/user.spec.js]: I just followed what Kalifat did when as she put the user controller in another file
For the user controller
I'll change the status code as soon as possible

File: server/routes/index.js (hunk 4 above)
Comment (cvjude) [reply server/routes/index.js:6]: The route doesn't start with auth
It starts with oauth
That's why I separated them

**Review by encodedBicoding — APPROVED** [review]

**Review by ocranbillions — APPROVED** [review]

**Review by chuxmykel — APPROVED** [review]

File: server/__test__/user.spec.js [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+
+    afterEach(() => {
+      authenticate.restore();
+    });
+    it('should signup a new user from google', done => {
+      authenticate.yields(null, users[1]);
+      server()
+        .post(`${url}/oauth/google`)
+        .end((err, res) => {
+          expect(res.statusCode).toEqual(200);
```
Comment (chuxmykel) [inline server/__test__/user.spec.js:80]: Is this supposed to be 200?

File: server/__test__/user.spec.js [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+          expect(res.body.message).toEqual('Login successful');
```
Comment (chuxmykel) [inline server/__test__/user.spec.js:81]: And is this Login or Signup?

File: server/__test__/user.spec.js [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+          expect(Object.keys(res.body.data)).toMatchSnapshot();
+        });
+    });
+
+    it('should signup a new user from facebook', done => {
+      authenticate.yields(null, users[2]);
+      server()
+        .post(`${url}/oauth/facebook`)
+        .end((err, res) => {
+          expect(res.statusCode).toEqual(200);
```
Comment (chuxmykel) [inline server/__test__/user.spec.js:108]: 200?

File: server/__test__/user.spec.js [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+          expect(res.body.message).toEqual('Login successful');
```
Comment (chuxmykel) [inline server/__test__/user.spec.js:109]: Login?

File: server/__test__/user.spec.js (hunk 5 above)
Comment (cvjude) [reply server/__test__/user.spec.js:80]: Well it's supposed to be a sign in that's the reason for the 200, if u don't have an account we subtlety add u to the database.

File: server/__test__/user.spec.js (hunk 6 above)
Comment (cvjude) [reply server/__test__/user.spec.js:81]: It's a login, but the user is not requires to sign up,  if u don't have an account we just add you

File: server/__test__/user.spec.js (hunk 8 above)
Comment (cvjude) [reply server/__test__/user.spec.js:109]: As I explained above

File: server/__test__/user.spec.js (hunk 6 above)
Comment (chuxmykel) [reply server/__test__/user.spec.js:81]: Airtight. Cool

File: server/__test__/user.spec.js (hunk 7 above)
Comment (encodedBicoding) [reply server/__test__/user.spec.js:108]: Has this been resolved?

**Review by encodedBicoding — APPROVED** [review]

**Pull request merged**